### PR TITLE
Transform POC tiles into TMS tiles and writhe to our TMS bucket

### PR DESCRIPTION
**Pull request by gjn**
This PR adds a utility which allows to

1) get defined tiles from POC (we know it works in POC)
2) do something with it
3) write them back to our bucket (to test if they work in POC)

Call it with `./venv/bin/python forge/scripts/poctiles2tmstiles.py forge/data/level8-all.txt`

The input is a text file defining the tile urls that you want to consider. The forge/data/level8-all.txt file contains all level 8 tiles covering whole of Switzerland. It's not that many and well visible in the POC. 

What was done with this utility.

# First run
1) read POC tile into TerrainTile
2) Created blob with toStringIO()
3) put gzipped blob it into tms bucket
-> tiles can be seen in POC

# Second run
1) read POC tile into TerrainTile
2) write new terrain file
3) read this new terrain file into a new TerrainTile
2) Created blob with toStringIO()
3) put gzipped blob it into tms bucket
-> tiles can be seen in POC

# Third run
1) read POC tile into TerrainTile
2) write terrainTile to shapeFile
3) read shapefile with topology helper class
4) created new terrain file with topology
2) Created blob with toStringIO()
3) put gzipped blob it into tms bucket
-> tiles can **NOT** be seen in POC

[As can be seen here](https://www.diffchecker.com/fy5wlvfp) (created with `8/267/194`), it seems like the headers are quite wrong.

I will now try to replace the header completely and leaving the coordinates as it is to see if it works.

### CODE REVIEW

**Comment by gjn** [comment]
Fourth run: same as third, but before writing the file, I replaced the header completely with the header from the original file -> tiles can still **NOT** be seen.

It might be that we have the triangles wrong (wrong order maybe?)

**Comment by gjn** [comment]
Fifth run: same as third, but I replace the header of the original file with the shape-file based header. Coordinates/triangles are therefore as in the original -> tilse can still **NOT** be seen.

There is a slight difference between runs [4] and [5]. In [4], in the POC the tile borders could be seen (when activating tile borders in debug widget). In [5], not even the tile borders are visible.Therefore, we can conclude that a) the header is important to show the borders. The coordinates/triangles to show everything.

Anyway, I would propose to use the fourth run as base and fix the coordinates first (use header from original, coordinates from our calculations). Once this is fully working, we can dive into correcting the header. Or we could separate the tasks among people (I will return on Tuesday)

ping @procrastinatio @loicgasser

**Comment by loicgasser** [comment]
Thanks for this intense testing. You really narrowed it down. I believe triangles order in index data is probably quite important.

Hypothesis 1:
Triangles defined in IndexData must be declared so that they are connected to one another (even though this is not mentioned anywhere in the doc). This could be checked by reading a tile from POC and by checking indexes 3 by 3. If after each iteration triangles have at least one coordinate in common, then it means we need to adapt the topology creation.

Hypothesis 2:
Vertices defined in edgeIndices must declared differently.
From doc:
Each triplet of indices specifies one triangle to be rendered, in counter-clockwise winding order.
:)
So this would mean we need to consider the vertices order as well...

**Comment by gjn** [comment]
I tend to think that winding order might be wrong for some triangles (hypo 2). The posted diff (link in OP, uvertex part) shows same coordinates, but some triples are in different order (2nd and 3rd coordinate reversed), supporting hypo 2. Seems like  shapefiles defines somemtriangles clockwise. Looks like we need to analyse each triplet and assure correct winding order.

The headers worry me more...

**Comment by loicgasser** [comment]
I ll fix 2 amd check for 1 on Monday. About header, the center of the tile is computed in 3d. I may instead need to compute the center of the tile in 2D and then interpilate the height instead. Not sure about that yet. Some thoughts?

**Comment by gjn** [comment]
I work today. I can do 2) and concentrate on the coordinates while you concentrate on the header?

I'll continue the analysis anyway.

**Comment by gjn** [comment]
I've now analysed the headers in more detail to find out which one fails. What I did was the following:

1. ReadPOC tile
2. Create shapefile, create new terrain file based on it
3. Use specific calculated header and apply it to the original POC tile
4. See if it works in the viewer

Headers that seem to be accepted by cesium:
- center (loads kind of slow, but they render well)
- min/max height
- boundingsphere

Header that break the tile rendering:
- occlusion point

It's clear to me that the occlusion point calculation needs to be made. @loicgasser Could you focus on that while I fix the coordinates?

(Other headers maybe need adaptions as well to be fully correct, but at least, they seem good enough for cesium to render the tiles)

**Comment by loicgasser** [comment]
ok I ll focus on occlusion point today.

**Comment by gjn** [comment]
I've added a commit now that reverses winding order from rings read in from shapefiles, eh voila, the generated tiles works when we adapt the occlusion point.

I will now try to use this code to create layer 8 and 9 with the dtm25 model we have and see if it works.

Would be good if you could review/merge this (we still need general purpose winding order check...see comments in code)

**Comment by loicgasser** [comment]
yes will do

**Comment by gjn** [comment]
Occlusion point calcuation: https://cesiumjs.org/2013/05/09/Computing-the-horizon-occlusion-point/

Shape files seem to be in clockwise order: http://gis.stackexchange.com/questions/48747/polygon-creation-clockwise-rotation-or-not

**Comment by gjn** [comment]
Ok, I've added the occlusion point calculation. It was rather easy by following the doc above.

Now, full roundtrip results in tiles that can be rendered by the application. Hurray.

File: forge/lib/topology.py [hunk 1]
```diff
@@ -89,7 +89,49 @@ def _ringFromGDALGeometry(self, geometry):
         # Remove last point of the polygon and keep only 3 coordinates
         return points[0: len(points) - 1]
 
+    def _isRingClockWise(self, ring):
+        '''
+        The below code does not work
+        correctly for sphere coordinates (lon/lat)
+        Therefore, I commente it out.
```
Comment (loicgasser) [inline forge/lib/topology.py]: commented or comment

File: forge/lib/horizon_occlusion_point.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
+# -*- coding: utf-8 -*-
+
+import math
+import forge.lib.cartesian3d as c3d
+import numpy as np
+
+# Constants taken from http://cesiumjs.org/2013/04/25/Horizon-culling/
+rX = 1.0 / 6378137.0
+rY = 1.0 / 6378137.0
+rZ = 1.0 / 6356752.3142451793
```
Comment (loicgasser) [inline forge/lib/horizon_occlusion_point.py]: we can avoid duplicating these values, they are already here https://github.com/geoadmin/3d-forge/blob/master/forge/lib/llh_ecef.py

Comment (gjn) [inline forge/lib/horizon_occlusion_point.py]: but different naming and different precision.

I agree to define them once.

**Comment by loicgasser** [comment]
Hourrah!

**Comment by loicgasser** [comment]
:tada:

File: forge/lib/topology.py [hunk 3]
```diff
@@ -89,7 +89,49 @@ def _ringFromGDALGeometry(self, geometry):
...
+            if index == len(ring) - 1:
+                p2 = ring[0]
+            else:
+                p2 = ring[index + 1]
+
+            area += ((p1[0] * p2[1]) - (p2[0] * p1[1]))
+
+        return area > 0.0
+        '''
+        return True
```
Comment (loicgasser) [inline forge/lib/topology.py]: ok for now. ;) Have a solution to check winding order.

**Comment by gjn** [comment]
I'm currently creating level 8 with code from this branch using mnt25 shapefiles....and it's working. But, compared to POC tiles, we have some artefacts when rendering...especially at the borders. See tiles on the left in below screenshot (tiles on the right are from POC):

image

File: forge/lib/topology.py [hunk 4]
```diff
@@ -89,7 +89,49 @@ def _ringFromGDALGeometry(self, geometry):
...
+
+    def _assureRingCounterClockWise(self, ring):
+        if len(ring) != 3:
+            raise TypeError('A ring must have exactly 3 coordinates.')
+
+        # If clockwise, we make it counterclockwise
+        if self._isRingClockWise(ring):
+            temp = ring[1]
```
Comment (loicgasser) [inline forge/lib/topology.py]: what about

ring[1], ring[2] = ring[2], ring[1]

Comment (gjn) [inline forge/lib/topology.py]: the python way :). I'll adapt.

**Comment by gjn** [comment]
I've adapted according to your comments. Would be good to merge, then we can start fine-tune on the same code base.

**Comment by loicgasser** [comment]
borders artefacts are probably there because vertices on edges don't match exactly as I explained earlier. During coordinates interpolation (or quantization), and in the next tile we quantize values again and as a result values don't match exactly... I belive this is the issue.

**Comment by gjn** [comment]
Maybe we can determine the border coordinates (westI, southI, northI, eastI) based on quantized coordinates (u, v) instead of full lon/lat (uVertex, vVertex). In u,v, we only need to check for 0 and MAX.

https://github.com/geoadmin/3d-forge/blob/master/forge/models/terrain.py#L435

**Comment by gjn** [comment]
I'm currently processing level 8 and 9 with multiprocessing. It's kind of slow and the speed heavily depends on the number of triangles we have after clipping. For level 8 and 9, we probably have way too many with mnt25.

For performance, it's important to select the correct model for each zoom level.

I let the 8/9 update run.

**Comment by gjn** [comment]
[10613] It took 0:06:00.481137 to create 8/269/194.terrain tile on S3. (1 total in this process)

It took over 6 minutes to create 8/269/194.terrain from mnt25...

**Comment by loicgasser** [comment]
yes ok but have you looked at the stats??

We have way to many triangles atz zoom level 8.

At zoom 8:
We expect 14 tiles overall
Min X is 264, Max X is 271
7 columns over X
Min Y is 193, Max Y is 195
2 rows over Y

A tile side is around 95242 meters long
We have an average of about 37869 triangles per tile

**Comment by loicgasser** [comment]
you have to start at level 9

**Comment by loicgasser** [comment]
can you come on gmail chat?

**Comment by gjn** [comment]
Some tiles on level 9 with mnt25 show the same artefacts (tile is outside of switzerland):

image

Regarding number of triangles: exactly. We need to define which model for which zoom level.

Multiprocessing is not stable enough. Looks to me that after some time, processes go stale and don't process any more tiles. Will investiagte.

**Comment by gjn** [comment]
hangout in afternoon.

**Comment by gjn** [comment]
How many tiles for level 9? 100?

**Comment by loicgasser** [comment]
make tmsstats

At zoom 9:
We expect 70 tiles overall
Min X is 528, Max X is 542
14 columns over X
Min Y is 386, Max Y is 391
5 rows over Y

A tile side is around 47672 meters long
We have an average of about 7573 triangles per tile

**Comment by gjn** [comment]
That's nice. Thanks alot.

When I run only level nine, 90 tiles were created (in less than 3 minutes). Why is there difference between actual tiles created and make tmsstats?

**Comment by loicgasser** [comment]
We should avoid creating all jobs in a list in the first place. If we can smaller batches or use functional programming we'd be much better off. 49'26'288 tiles at zoom level 17 at the moment not really sustainable with current multiprocessing.

**Comment by loicgasser** [comment]
Just generated 9 to 12 at zoom took something like 7 minutes.

**Comment by gjn** [comment]
Sure.

Anyway, the 9 level tiles, most still can't be displayed with renderer. So there, troubles in terrain remain.

**Comment by loicgasser** [comment]
I can see all of them just change your apache conf

**Comment by gjn** [comment]
? my apache conf uses our tiles for all zoom leves >= 8

**Comment by gjn** [comment]
RewriteRule ^/ltjeg/tiles/(8|9|[0-9]{2})/(.*) http://tms3d.geo.admin.ch.s3.amazonaws.com/$1/$2 [L,P,QSA]

**Comment by loicgasser** [comment]
/ltjeg/tiles/(9|10|11|12)/  no?

**Comment by gjn** [comment]
It's the same. mine includes all 2 digit zoom levels.

**Comment by gjn** [comment]
In any case, rendering is kind of slow (too many triangles) and with lots of artefacts.

image

There's still work to do :=)

**Comment by loicgasser** [comment]
zoom_9

**Comment by loicgasser** [comment]
8 is too big to be rendered, maybe that's your problem.

**Comment by gjn** [comment]
I'll replace level 8 with the POC tiles again.

**Comment by gjn** [comment]
done. Let's talk in the afternoon (chat)

**Comment by gjn** [comment]
Probably the artefacts are the clockwise stuff. I think in topology, when you do the real clipping (collapse geometry), you probably create anti-clockwise triangles, which will then reversed in the terrain class.

We need real winding direction test.

**Comment by loicgasser** [comment]
yes looking at the code https://github.com/geoadmin/3d-forge/blob/master/forge/models/terrain.py#L413

bLon and bLat should be the same for the same zoom level.
When do you want to talk?

**Comment by loicgasser** [comment]
I mean what time?

**Pull request merged**